[3D] Add transparent support to the datadefined case of phong material

ptitjano opened the pull request:
An initial transparency support was added to phong material in https://github.com/qgis/QGIS/pull/48499. However, it did not handle the datadefined case. This PR fills the gap by adding the `opacity` value to the different color attributes. 

without opacity
![pas d'opacité](https://user-images.githubusercontent.com/497207/178275665-8c16182f-7ded-4171-9085-3d7cd11beeec.png)

50% opacity
![opacité 50%](https://user-images.githubusercontent.com/497207/178275725-065a2781-9047-4001-878b-cf5785470da4.png)

cc @NEDJIMAbelgacem @wonder-sk @benoitdm-oslandia

ptitjano committed:
qgsphongmaterialsettings: Add opacity support to datadefined

The opacity parameter is added to the effect used by phongDataDefined
shader. It is then used in the vertex shader instead of the default
1.0 value.

As far as the FrameGraph is concerned, the transparent layer needs to
be added to the corresponding Qt3DCore::QEntity. This is handled in
Qgs3DMapScene::finalizeNewEntity by checking all the
material which have an opacity parameter.

ptitjano committed:
qgsphongmaterialsettings: Fix opeator== by adding mOpacity parameter

ptitjano committed:
qgsphongmaterialsettings: Fix setOpacity documentation

ptitjano committed:
qgs3dmapscene: Remove empty line at the end of file

ptitjano committed:
phongdatadefined: Remove empty line at the end of file

benoitdm-oslandia reviewed (approved):
Thanks @ptitjano

wonder-sk commented on src/3d/shaders/phongDataDefined.vert:
```diff
@@ -33,11 +33,10 @@ void main()
-    vs_out.specular=vec4(dataDefinedSpecularColor,1.0);
+    vs_out.ambient=dataDefinedAmbiantColor;
+    vs_out.diffuse=dataDefinedDiffuseColor;
+    vs_out.specular=dataDefinedSpecularColor;
```
Are you sure this is correct to apply opacity to all the colors here? Isn't that going to end up with incorrect color mixing? Shouldn't we apply opacity to the final color at the very end of the fragment shader?

ptitjano replied:
In fact, `phongFunction` only takes into account the rgb components of the colors and then applies the opacity of the diffuse color to the final color. Therefore, there is no risk of incorrect color mixing, this is already handled.

This means, that one only need to change `vs_out.diffuse` to have the expected result. However, it seems odd (to me) to change only one color. What do you prefer?

ptitjano committed:
qgsphongmaterialsettings: Directly add shader parameters to the effect

Adding a parameter to a technique is only useful when different
techniques are used.
This is also the behavior used by addParametersToEffect which directly
adds parameters to the effect.

ptitjano commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -928,16 +928,45 @@ void Qgs3DMapScene::finalizeNewEntity( Qt3DCore::QEntity *newEntity )
-    if ( entity->components().contains( layer ) )
-      continue;
-    entity->addComponent( layer );
+    // this handles the untextuted case without data defined properties
```
fixed.

ptitjano commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -928,16 +928,45 @@ void Qgs3DMapScene::finalizeNewEntity( Qt3DCore::QEntity *newEntity )
+      {
+        if ( !newEntity->components().contains( layer ) )
+          newEntity->addComponent( layer );
+      }
```
You're right. I had completely failed to take into account the subentity logic. This should be fixed now.

ptitjano commented on src/3d/materials/qgsphongmaterialsettings.cpp:
```diff
@@ -160,25 +161,29 @@ QByteArray QgsPhongMaterialSettings::dataDefinedVertexColorsAsByte( const QgsExp
   *fptr++ =  static_cast<unsigned char>( specular.red() );
   *fptr++ =  static_cast<unsigned char>( specular.green() );
   *fptr++ =  static_cast<unsigned char>( specular.blue() );
+  *fptr++ = opacity;
```
With this new version, the opacity parameter is now added as a uniform parameter to the effect.

wonder-sk commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -924,20 +924,43 @@ void Qgs3DMapScene::finalizeNewEntity( Qt3DCore::QEntity *newEntity )
+            }
+            break;
+          }
+        }
```
As the code starts to get more complex, ideally we should wrap the whole content QgsPhongMaterialSettings::dataDefinedMaterial() into a new class (e.g. QgsDataDefinedPhongMaterial), and then we could handle this case just by casting to this material class like with QDiffuseSpecularMaterial (to avoid going through effects and their parameters). But probably that's something that can be done in a future PR...

ptitjano replied:
I have to admit, I don't see much added value at the moment.

wonder-sk reviewed (commented):
Thanks for the updates, it's looking better!

ptitjano commented on src/3d/materials/qgsphongmaterialsettings.cpp:
```diff
@@ -222,7 +227,11 @@ void QgsPhongMaterialSettings::applyDataDefinedToGeometry( Qt3DRender::QGeometry
+  if ( mOpacity != 1.0 )
+  {
+    material->setAlphaBlendingEnabled( true );
+  }
```
switched to `QMaterial`.

wonder-sk reviewed (approved):
Looks good to me, thanks!
